finance-api/internal/service/invoices.go

```
@@ -43,18 +27,58 @@ func (s *Service) GetInvoices(clientID int) (*shared.Invoices, error) {
 			RaisedDate:         shared.Date{Time: i.Raiseddate.Time},
 			Received:           totalOfLedgerAllocationsAmount,
 			OutstandingBalance: int(i.Amount) - totalOfLedgerAllocationsAmount,
-			Ledgers:            ledgerAllocations,
-			SupervisionLevels:  nil,
+			Ledgers:            getLedgerAllocations(s, ctx, int(i.ID)),
+			SupervisionLevels:  getSupervisionLevels(s, ctx, int(i.ID)),
 		}
 		invoices = append(invoices, invoice)
 	}
 
 	return &invoices, nil
 }
 
-func CheckForAllocation(allocatedDate pgtype.Date, datetime pgtype.Timestamp) shared.Date {
-	if !allocatedDate.Time.IsZero() {
-		return shared.Date{Time: allocatedDate.Time}
+func getSupervisionLevels(s *Service, ctx context.Context, invoiceID int) []shared.SupervisionLevel {
+	var supervisionLevels []shared.SupervisionLevel
+	supervisionLevelsRawData, err := s.Store.GetSupervisionLevels(ctx, pgtype.Int4{Int32: int32(invoiceID), Valid: true})
+	if err != nil {
+		return nil
+	}
+
+	for _, i := range supervisionLevelsRawData {
+		var supervisionLevel = shared.SupervisionLevel{
+			Level:  i.Supervisionlevel,
+			Amount: int(i.Amount),
+			From:   shared.Date{Time: i.Fromdate.Time},
+			To:     shared.Date{Time: i.Todate.Time},
+		}
+		supervisionLevels = append(supervisionLevels, supervisionLevel)
+	}
+	return supervisionLevels
+}
+
+func getLedgerAllocations(s *Service, ctx context.Context, invoiceID int) []shared.Ledger {
+	var ledgerAllocations []shared.Ledger
+	totalOfLedgerAllocationsAmount = 0
+	ledgerAllocationsRawData, err := s.Store.GetLedgerAllocations(ctx, pgtype.Int4{Int32: int32(invoiceID), Valid: true})
+	if err != nil {
+		return nil
```

Shouldn't this return an error?

---

I am not sure how I would be able to return an error as the invoice struct for ledgers only expect an array of allocations, happy to talk about it tho

---

If the query is incorrect or returns invalid data, this would silently fail and we would have no indication it was wrong until it failed for users in production. If the code cannot return an error, it has to be refactored so it can. We can't just ignore errors because they are hard to write.

Functions in Go can return multiple types. The `Store.GetLedgerAllocations` function two lines above does exactly that.